crates/uv-resolver/src/resolver/derivation.rs

```diff
@@ -0,0 +1,137 @@
+use std::collections::VecDeque;
+
+use petgraph::Direction;
+use pubgrub::{Kind, SelectedDependencies, State};
+use rustc_hash::FxHashSet;
+
+use uv_distribution_types::{
+    DerivationChain, DerivationStep, DistRef, Name, Node, Resolution, ResolvedDist,
+};
+use uv_pep440::Version;
+
+use crate::dependency_provider::UvDependencyProvider;
+use crate::pubgrub::PubGrubPackage;
+
+/// A chain of derivation steps from the root package to the current package, to explain why a
+/// package is included in the resolution.
+#[derive(Debug, Default, Clone, PartialEq, Eq, Hash)]
+pub struct DerivationChainBuilder;
+
+impl DerivationChainBuilder {
+    /// Compute a [`DerivationChain`] from a resolution graph.
+    ///
+    /// This is used to construct a derivation chain upon install failure in the `uv pip` context,
+    /// where we don't have a lockfile describing the resolution.
+    pub fn from_resolution(
+        resolution: &Resolution,
+        target: DistRef<'_>,
+    ) -> Option<DerivationChain> {
+        // Find the target distribution in the resolution graph.
+        let target = resolution.graph().node_indices().find(|node| {
+            let Node::Dist {
+                dist: ResolvedDist::Installable { dist, .. },
+                ..
+            } = &resolution.graph()[*node]
+            else {
+                return false;
+            };
+            target == dist.as_ref()
+        })?;
+
+        // Perform a BFS to find the shortest path to the root.
+        let mut queue = VecDeque::new();
+        queue.push_back((target, Vec::new()));
+
+        // TODO(charlie): Consider respecting markers here.
```

Comment:
Without markers, I think this means that you might find the shortest path, but that path might not be possible? Or is there another downside I'm not seeing?